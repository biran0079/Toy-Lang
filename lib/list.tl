fun prefix(l, n){
  res = [];
  i=0;
  while(i<n){
    res += l[i];
    i += 1;
  }
  return res;
}
fun suffix(l,n){
  res=[];
  length = len(l);
  i=n;
  while(i<length){
    res += l[i];
    i += 1;
  }
  return res;
}


fun swap(a,i,j){
  t=a[i];
  a[i]=a[j];
  a[j]=t;
}

fun isSorted(l){
  length = len(l);
  i=1;
  while(i<length){
    if(l[i-1] > l[i])
      return 0;
    i=i+1;  
  }
  return 1;
}
