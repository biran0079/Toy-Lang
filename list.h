#ifndef _LIST_H_
#define _LIST_H_

typedef struct List {
  void** arr;
  int size, cap;
} List;

List* newList();
void listPush(List* lst, void* e);
void* listGet(List* lst, int idx);
void* listPop(List* lst); 
int listSize(List* lst);
void* listLast(List* lst);
void listSet(List* lst, int i, void* v);
List* listCopy(List* lst);
void listClear(List* lst);

#endif
