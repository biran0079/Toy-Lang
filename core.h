#ifndef _CORE_H_
#define _CORE_H_
#include "list.h"

typedef struct Token Token;
typedef struct Node Node;
typedef struct Value Value;
typedef struct Closure Closure;
typedef struct Env Env;
typedef enum ValueType ValueType;
typedef enum NodeType NodeType;
typedef void (*ValueFunc)(Value*);
typedef Value* (*BuiltinFun)(List* l); 

void init(int argc, char** argv);
void cleanup();
long getIntId(char *s);
char* getStrId(long id);

extern int newNodeC, freeNodeC;
extern int newListC, freeListC;
extern int newHashTableC, freeHashTableC;
extern int newIntValueC, newStringValueC, newClosureValueC, newEnvValueC, newListValueC, newBuiltinFunC;
extern int freeIntValueC, freeStringValueC, freeClosureValueC, freeEnvValueC, freeListValueC, freeBuiltinFunC;
extern int newClosureC, freeClosureC;
extern int newEnvC, freeEnvC;
extern int memoryUsage;

#define YYSTYPE Node*

#endif
