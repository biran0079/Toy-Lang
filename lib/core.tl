import hashtable;
import list;

fun newEnv(p){
  local res;
  res = [p,hashtable.newStringHashTable()];
  envPut(res, "this", res);
  return res;
}

fun envHas(e, key) {
  while(e!=none){
    if(hashtable.has(e[1], key)) {
      return 1;
    }
    e=e[0];
  }  
  return 0;
}

fun envGet(e, key) {
  while(e!=none){
    if(hashtable.has(e[1], key)) {
      return hashtable.get(e[1],key);
    }
    e=e[0];
  }  
  return none;
}

fun envPut(e, key, value) {
  e0 = e;
  while(1){
    if(hashtable.has(e[1], key)){
      hashtable.put(e[1], key, value);
      return;
    } else {
      if(e[0] == none)  break;
      e = e[0];
    }
  }
  hashtable.put(e0[1], key, value);
}

fun envPutLocal(e, key, value) {
  hashtable.put(e[1], key, value);
}


fun assert(b, msg){
  if(!b){
    print("assertion failure\n"); 
    print(msg);
    exit(-1);
  }
}

fun setUpBuiltinFuns(e){
  envPut(e, "print", newBuiltinFun(print));
  envPut(e, "len", newBuiltinFun(len));
  envPut(e, "ord", newBuiltinFun(ord));
  envPut(e, "chr", newBuiltinFun(chr));
  envPut(e, "sort", newBuiltinFun(sort));
  envPut(e, "str", newBuiltinFun(str));
  envPut(e, "rand", newBuiltinFun(rand));
  envPut(e, "parse", newBuiltinFun(parse));
  envPut(e, "read", newBuiltinFun(read));
  envPut(e, "exit", newBuiltinFun(exit));
  envPut(e, "apply", builtinApply);
  envPut(e, "sysArgs", newBuiltinFun(lambda() {return list.suffix(sysArgs(), 1);}));
  return res;
}

path = ["lib/", "./"];
fun readFromPath(f){
  local i;
  for(i=0;i<len(path);i++){
    s = read(path[i]+f);
    if(s!=none) return s;
  }
  print("failed to read " + f + "from path");
  exit(-1);
}

fun newGlobalEnv() {
  local res;
  res = newEnv(none);
  setUpBuiltinFuns(res);
  return res;
}

fun newBuiltinFun(f) {
  fun internalBuiltinFun(args) {
    throw ["return", apply(f, args)];  
  }
  return internalBuiltinFun;
}

fun builtinApply(args) {
  throw ["return", apply(args[0], list.suffix(args, 1))];
}

fun newInternalFun(e, ids, fbody) {
  fun internalFunction(args) {
    local e2, i;
    e2 = newEnv(e);
    assert(len(ids) == len(args),str(len(ids)) + " parameter needed, " + str(len(args)) + " parameters get\n");
    for(i=0;i<len(ids);i++)
      envPutLocal(e2, ids[i][1], args[i]);
    throw ["return", eval(e2, fbody)];
  }
  return internalFunction;
}
