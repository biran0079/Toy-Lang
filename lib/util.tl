fun getRandomList(n){
  res=[];
  while(n>0){
    res += rand();  
    n = n - 1;
  }
  return res;
}


import list;

fun internalQsort(a, left, right){
  if(left>=right)return;
  i=left;
  j=left+1;
  k=j;
  while(j<=right){
    if(a[j]<a[i]) {
      list.swap(a,k,j);
      k=k+1;
    }
    j=j+1;  
  }
  k=k-1;
  list.swap(a, left, k);
  internalQsort(a,left,k-1);
  internalQsort(a,k+1,right);
}
fun qsort(l){
  internalQsort(l,0,len(l)-1);  
}


fun merge(l1, l2){
  res=[];
  len1 = len(l1);
  len2 = len(l2);
  i=0;
  j=0;
  while(i<len1 && j<len2){
    if(l1[i]<l2[j]){
      res += l1[i];  
      i += 1;
    }else{
      res += l2[j];
      j = j+1;
    }
  }
  while(i<len1) {
    res += l1[i];  
    i += 1;
  }
  while(j<len2) {
    res += l2[j];
    j += 1;
  }
  return res;
}
fun msort(lst){
  length = len(lst);
  if(length<=1)return;
  mid = length / 2;
  l1 = list.prefix(lst, mid);
  l2 = list.suffix(lst, mid);
  msort(l1);
  msort(l2);
  t = merge(l1, l2);
  i=0;
  while(i<length){
    lst[i] = t[i];
    i=i+1;  
  }
}

