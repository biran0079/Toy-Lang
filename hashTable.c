#include "hashTable.h"
#include "list.h"
#include "util.h"

int newHashTableC = 0, freeHashTableC = 0;

static unsigned int stringHash(unsigned char *s) {
  unsigned int h = 0;
  while (*s) {
    h *= 127;
    h += *s;
    s++;
  }
  return h;
}

static int stringEqual(char *s1, char *s2) { return strcmp(s1, s2) == 0; }

static unsigned int intHash(long n) { return (unsigned int)n; }

static int intEqual(long a, long b) { return a == b; }

static int INIT_CAPACITY = 8;

extern int newHashTableC, freeHashTableC;

HashTable *newStringHashTable() {
  return newHashTable(stringHash, stringEqual);
}

HashTable *newIntHashTable() { return newHashTable(intHash, intEqual); }

HashTable *newHashTable(HashFunc h, EqualsFunc eq) {
  newHashTableC++;
  HashTable *res = MALLOC(HashTable);
  res->cap = INIT_CAPACITY;
  res->size = 0;
  res->a = (LinkedList **)tlMalloc(res->cap * sizeof(LinkedList *));
  res->h = h;
  res->eq = eq;
  memset(res->a, 0, res->cap * sizeof(LinkedList *));
  return res;
}

static void rehash(HashTable *t) {
  LinkedList **old = t->a;
  int oldCap = t->cap;
  t->cap *= 2;
  t->a = (LinkedList **)tlMalloc(t->cap * sizeof(LinkedList *));
  t->size = 0;
  memset(t->a, 0, t->cap * sizeof(LinkedList *));
  int i;
  for (i = 0; i < oldCap; i++) {
    LinkedList *l = old[i], *next;
    while (l) {
      hashTablePut(t, l->key, l->value);
      next = l->next;
      l->next = 0;
      tlFree(l);
      l = next;
    }
  }
  tlFree(old);
}

void freeHashTable(HashTable *t) {
  freeHashTableC++;
  int i;
  for (i = 0; i < t->cap; i++) {
    LinkedList *l = t->a[i];
    while (l) {
      LinkedList *nl = l->next;
      l->key = 0;
      l->value = 0;
      l->next = 0;
      tlFree(l);
      l = nl;
    }
  }
  tlFree(t->a);
  t->a = 0;
  t->cap = t->size = 0;
  tlFree(t);
}

static LinkedList *hashTableGetInternal(HashTable *t, void *key) {
  unsigned idx = t->h(key) % t->cap;
  LinkedList *l = t->a[idx];
  while (l) {
    if (t->eq(l->key, key)) return l;
    l = l->next;
  }
  return 0;
}

void *hashTableRemove(HashTable *t, void *key) {
  unsigned idx = t->h(key) % t->cap;
  LinkedList *l = t->a[idx], *p = 0;
  while (l) {
    if (t->eq(l->key, key)) break;
    p = l;
    l = l->next;
  }
  if (l == 0) return 0;
  void *res = l->value;
  if (!p) {
    t->a[idx] = l->next;
    tlFree(l);
  } else {
    p->next = l->next;
    tlFree(l);
  }
  return res;
}

void hashTablePut(HashTable *t, void *key, void *value) {
  if (t->size > t->cap) rehash(t);
  LinkedList *l = hashTableGetInternal(t, key);
  if (l) {
    l->value = value;
  } else {
    unsigned int idx = t->h(key) % t->cap;
    l = MALLOC(LinkedList);
    l->key = key;
    l->value = value;
    l->next = t->a[idx];
    t->a[idx] = l;
    t->size++;
  }
}

void *hashTableGet(HashTable *t, void *key) {
  LinkedList *l = hashTableGetInternal(t, key);
  return l ? l->value : 0;
}

HashTable *hashTableCopy(HashTable *t) {
  HashTable *res = newHashTable(t->h, t->eq);
  int i;
  for (i = 0; i < t->cap; i++) {
    LinkedList *l = t->a[i];
    while (l) {
      hashTablePut(res, l->key, l->value);
      l = l->next;
    }
  }
  return res;
}

void hashTableClear(HashTable *t) {
  int i;
  for (i = 0; i < t->cap; i++) {
    LinkedList *l = t->a[i], *nl;
    while (l) {
      nl = l->next;
      tlFree(l);
      l = nl;
    }
    t->a[i] = 0;
  }
  t->size = 0;
}

void hashTableApplyAllValue(HashTable *t, ValueFunc f) {
  int i;
  for (i = 0; i < t->cap; i++) {
    LinkedList *l = t->a[i], *nl;
    while (l) {
      nl = l->next;
      f(l->value);
      l = nl;
    }
  }
}

void hashTableAddAllToList(HashTable *t, List *q) {
  int i;
  for (i = 0; i < t->cap; i++) {
    LinkedList *l = t->a[i], *nl;
    while (l) {
      nl = l->next;
      listPush(q, l->value);
      l = nl;
    }
  }
}

void freeStringHashTable(HashTable *t) {
  int i;
  for (i = 0; i < t->cap; i++) {
    LinkedList *l = t->a[i];
    while (l) {
      tlFree(l->key);
      l = l->next;
    }
  }
  freeHashTable(t);
}
