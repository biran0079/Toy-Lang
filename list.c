#include "list.h"
#include "util.h"

static void resize(List *lst, int cap) {
  lst->arr = tlRealloc(lst->arr, cap * sizeof(void *));
  lst->cap = cap;
}

int newListC = 0, freeListC = 0;

#define INIT_LIST_CAP 1

List *newList() {
  newListC++;
  List *res = MALLOC(List);
  res->size = 0;
  res->cap = INIT_LIST_CAP;
  res->arr = tlMalloc(res->cap * sizeof(void *));
  return res;
}

void freeList(List *lst) {
  freeListC++;
  tlFree(lst->arr);
  lst->arr = 0;
  lst->cap = 0;
  lst->size = 0;
  tlFree(lst);
}

void listPush(List *lst, void *e) {
  if (lst->cap == lst->size) {
    resize(lst, lst->cap * 2);
  }
  lst->arr[lst->size++] = e;
}

void listPushFront(List *lst, void *e) {
  if (lst->cap == lst->size) {
    resize(lst, lst->cap * 2);
  }
  int i;
  for (i = lst->size; i > 0; i--) {
    lst->arr[i] = lst->arr[i - 1];
  }
  lst->arr[0] = e;
  lst->size++;
}

void *listPop(List *lst) {
  if (lst->size == 0) {
    error("cannot pop empty list\n");
  }
  return lst->arr[--lst->size];
}

void *listLast(List *lst) {
  assert(lst->size > 0);
  return lst->arr[lst->size - 1];
}

void *listSet(List *lst, int idx, void *v) {
  assert(idx < lst->size && idx >= 0);
  void *oldValue = lst->arr[idx];
  lst->arr[idx] = v;
  return oldValue;
}

List *listCopy(List *lst) {
  List *res = newList();
  int i, len = lst->size;
  for (i = 0; i < len; i++) listPush(res, listGet(lst, i));
  return res;
}

void listClear(List *lst) { lst->size = 0; }

static void msortInternal(void *a[], int l, int r, Comparator cmp, void *t[]) {
  if (r - l <= 1) return;
  int mid = (l + r) / 2;
  msortInternal(a, l, mid, cmp, t);
  msortInternal(a, mid, r, cmp, t);
  int k = l, i = l, j = mid;
  while (i < mid && j < r) {
    if (cmp(a[i], a[j]) <= 0)
      t[k++] = a[i++];
    else
      t[k++] = a[j++];
  }
  while (i < mid) t[k++] = a[i++];
  while (j < r) t[k++] = a[j++];
  for (i = l; i < r; i++) a[i] = t[i];
}

void listSort(List *lst, Comparator cmp) {
  void **t = tlMalloc(listSize(lst) * sizeof(void *));
  msortInternal(lst->arr, 0, lst->size, cmp, t);
  tlFree(t);
}

void listPopTo(List *lst, int size) {
  if (lst->size < size)
    error("size passed to listPopTo is less than list size\n");
  lst->size = size;
}

void increaseSizeTo(List *lst, int size) {
  if (size <= lst->size) return;
  int cap = lst->cap;
  if (cap < size) {
    while (cap < size) cap *= 2;
    resize(lst, cap);
  }
  memset(lst->arr + lst->size, 0, sizeof(void *) * (size - lst->size));
  lst->size = size;
}

void listSwap(List *lst, int i, int j) {
  if (i == j) return;
  void *t = listGet(lst, i);
  listSet(lst, i, listGet(lst, j));
  listSet(lst, j, t);
}
