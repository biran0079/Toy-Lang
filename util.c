#include "util.h"

void* copy(void* t, int size) {
  void* res = malloc(size);
  memcpy(res, t, size);
  return res;
}

char* copyStr(char* t) {
  return copy(t, strlen(t) + 1);
}

long strToLong(char* s){
  long res = 0;
  while(*s) {
    res*=10;
    res+=*s-'0';
    s++;
  }
  return res;
}

char* literalStringToString(char *s) {
  char *ss = malloc(strlen(s)+1);
  int l = 0;
  s++;  // skil "
  while(*s!='"') {
    if(*s == '\\') {
      s++;
      switch(*s){
        case 'n' : ss[l++] = '\n'; break;
        case 'r' : ss[l++] = '\r'; break;
        case '"' : ss[l++] = '"'; break;
        case '\\' : ss[l++] = '\\'; break;
        default:  ss[l++] = *s; break;
      }
    }else{
      ss[l++] = *s;
    }
    s++;
  }
  ss[l++] = 0;
  return realloc(ss, l);
}

void error(char* format,...) {
  va_list ap;
  va_start(ap, format);
  vfprintf(stderr, format, ap);
  va_end(ap);
  exit(-1);
}
