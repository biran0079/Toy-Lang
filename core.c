#include "ast.h"
#include "mem.h"
#include "core.h"
#include "env.h"
#include "builtinFun.h"
#include "gc.h"
#include "hashTable.h"
#include "value.h"
#include "util.h"
#include "eval.h"
#include "idMap.h"
#include "opStack.h"

List *parseTrees;  // make sure no tree is deleted during evaluation by keeping
                   // track of all parse trees
Env *globalEnv;
int memoryLimit = 64 * 1024 * 1024;
List *path;  // where import loads from
int isInitialized = 0;
int sysArgc;
char **sysArgv;

/*** ALL GLOBAL VARIABLES DECLARES ABOVE!  ***/

void listCreatedObjectsCount() {
  fprintf(stderr, "\tNode: %d (%d/%d)\n", newNodeC - freeNodeC, newNodeC,
          freeNodeC);
  fprintf(stderr, "\tIntValue: %d (%d/%d)\n", newIntValueC - freeIntValueC,
          newIntValueC, freeIntValueC);
  fprintf(stderr, "\tStringValue: %d (%d/%d)\n",
          newStringValueC - freeStringValueC, newStringValueC,
          freeStringValueC);
  fprintf(stderr, "\tClosureValue: %d (%d/%d)\n",
          newClosureValueC - freeClosureValueC, newClosureValueC,
          freeClosureValueC);
  fprintf(stderr, "\tListValue: %d (%d/%d)\n", newListValueC - freeListValueC,
          newListValueC, freeListValueC);
  fprintf(stderr, "\tClosure: %d (%d/%d)\n", newClosureC - freeClosureC,
          newClosureC, freeClosureC);
  fprintf(stderr, "\tEnv: %d (%d/%d)\n", newEnvC - freeEnvC, newEnvC, freeEnvC);
  fprintf(stderr, "\tList: %d (%d/%d)\n", newListC - freeListC, newListC,
          freeListC);
  fprintf(stderr, "\tHashTable: %d (%d/%d)\n", newHashTableC - freeHashTableC,
          newHashTableC, freeHashTableC);
  fprintf(stderr, "\tBuiltinFun: %d (%d/%d)\n",
          newBuiltinFunC - freeBuiltinFunC, newBuiltinFunC, freeBuiltinFunC);
  fprintf(stderr, "\tEvalResult: %d (%d/%d)\n",
          newEvalResultC - freeEvalResultC, newEvalResultC, freeEvalResultC);
  fprintf(stderr, "memory usage: %d\n", memoryUsage);
}

static void initPath(int argc, char** args) {
  path = newList();
  char *tlDir = getFolder(args[0]);
  listPush(path, catStr(tlDir, "lib/"));
  tlFree(tlDir);
  listPush(path, copyStr("./"));
  char *srcDir = getFolder(args[1]);
  if (strcmp(srcDir, "./")) {
    listPush(path, srcDir);
  }
}

static void cleanupPath() {
  int i, n = listSize(path);
  for (i = 0; i < n; i++) {
    tlFree(listGet(path, i));
  }
  freeList(path);
}

void init(int argc, char **args) {
  initMem();
  initPath(argc, args);
  initIdMap();
  initOpStack();

  // global env depends on value block (none value), id map (default 'this'
  // field)
  globalEnv = getEnvFromValue(newEnvValue(0));
  opStackPush(globalEnv->envValue);
  registerBuiltinFunctions(globalEnv);

  parseTrees = newList();

  isInitialized = 1;
}

void cleanup() {
  assert(globalEnv->envValue == opStackPop());  // make sure global env get GCed
  forceGC();
  cleanupOpStack();
  cleanupIdMap();
  cleanupPath();
  cleanupMem();
  int i, n = listSize(parseTrees);
  for (i = 0; i < n; i++) {
    freeNode(listGet(parseTrees, i));
  }
  isInitialized = 0;
}

