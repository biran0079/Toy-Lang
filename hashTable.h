#ifndef _HASH_TABLE_H_
#define _HASH_TABLE_H_

#include "core.h"
#include "list.h"

typedef struct LinkedList LinkedList;
struct LinkedList {
  void *key;
  void *value;
  LinkedList *next;
};

typedef unsigned int (*HashFunc)();
typedef int (*EqualsFunc)();

typedef struct HashTable {
  unsigned int cap, size;
  LinkedList **a;
  HashFunc h;
  EqualsFunc eq;
} HashTable;

HashTable *newHashTable(HashFunc h, EqualsFunc eq);
void freeHashTable(HashTable *t);
void *hashTablePut(HashTable *t, void *key, void *value);
void *hashTableGet(HashTable *t, void *key);
void *hashTableRemove(HashTable *t, void *key);
void hashTableClear(HashTable *t);
int hashTableSize(HashTable *t);

HashTable *newStringHashTable();
HashTable *newIntHashTable();
void freeStringHashTable(HashTable *t);
List *hashTableGetAllKeys(HashTable *t);
#endif
