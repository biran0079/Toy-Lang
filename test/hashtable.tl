fun newHashTable(){
  a = [];
  for(i=0;i<89;i+=1) 
    a+=[[]];
  return a;
}
fun hash(s){
  res = 0;  
  l=len(s);
  for(i=0;i<l;i+=1){
    res=res*89;
    res+=ord(s[i]);
  }
  return res;
}
fun hashTablePut(t, k, v){
  size = len(t);
  i = hash(k) % size;
  if(i<0) i += size;
  l = t[i];
  llen = len(l);
  for(i=0;i<llen;i+=1){
    if(l[i][0] == k) {
      l[i][1]=v;  
      return;
    }
  }
  l+=[[k,v]];
}

fun hashTableGet(t, k){
  size = len(t);
  i = hash(k) % size;
  if(i<0) i += size;
  l = t[i];
  llen = len(l);
  for(i=0;i<llen;i+=1)
    if(l[i][0] == k) 
      return l[i][1];
  return;
}

t = newHashTable();
hashTablePut(t, "hello", "world");
print(hashTableGet(t,"hello"));
hashTablePut(t, "hello", "toy-lang");
print(hashTableGet(t,"hello"));
hashTablePut(t, "123", "321");
print(hashTableGet(t,"123"));
print(hashTableGet(t,"333"));
print(hashTableGet(t,"333")==none);
