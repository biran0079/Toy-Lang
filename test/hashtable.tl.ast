(stmts (import (id hashtable)) (= (id HT) (id hashtable)) (= (id t) (call (module_access (id HT) (id newStringHashTable)) (exps))) (call (module_access (id HT) (id put)) (exps (id t) (string "hello") (string "world"))) (call (id print) (exps (call (module_access (id HT) (id get)) (exps (id t) (string "hello"))))) (call (module_access (id HT) (id put)) (exps (id t) (string "hello") (string "toy-lang"))) (call (id print) (exps (call (module_access (id HT) (id get)) (exps (id t) (string "hello"))))) (call (module_access (id HT) (id put)) (exps (id t) (string "123") (string "321"))) (call (id print) (exps (call (module_access (id HT) (id get)) (exps (id t) (string "123"))))) (call (id print) (exps (call (module_access (id HT) (id get)) (exps (id t) (string "333"))))) (call (id print) (exps (== (call (module_access (id HT) (id get)) (exps (id t) (string "333"))) (none)))))
