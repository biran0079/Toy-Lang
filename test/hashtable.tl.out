world
toy-lang
321
none
1
